core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,249 @@
+        };
+        Ok(Some((miniblock_number, l1_batch_number)))
+    }
+
+    /// Returns a [`ReadStorage`] implementation backed by Postgres
+    async fn access_storage_pg(
+        rt_handle: Handle,
+        pool: &ConnectionPool,
+    ) -> anyhow::Result<PgOrRocksdbStorage<'_>> {
+        let mut connection = pool.access_storage().await?;
+
+        let (miniblock_number, l1_batch_number) =
+            match Self::load_latest_sealed_miniblock(&mut connection).await? {
+                Some((miniblock_number, l1_batch_number)) => (miniblock_number, l1_batch_number),
+                None => {
+                    tracing::info!("Could not find latest sealed miniblock, loading from snapshot");
+                    let snapshot_recovery = connection
+                        .snapshot_recovery_dal()
+                        .get_applied_snapshot_status()
+                        .await
+                        .context("failed getting snapshot recovery info")?
+                        .ok_or_else(|| {
+                            anyhow::anyhow!("Could not find snapshot, no state available")
+                        })?;
+                    (
+                        snapshot_recovery.miniblock_number,
+                        snapshot_recovery.l1_batch_number,
+                    )
+                }
+            };
+
+        tracing::debug!(%l1_batch_number, %miniblock_number, "Using Postgres-based storage");
+        Ok(
+            PostgresStorage::new_async(rt_handle, connection, miniblock_number, true)
+                .await?
+                .into(),
+        )
+    }
+
+    /// Catches up RocksDB synchronously (i.e. assumes the gap is small) and
+    /// returns a [`ReadStorage`] implementation backed by caught-up RocksDB.
+    async fn access_storage_rocksdb<'a>(
+        conn: &mut StorageProcessor<'_>,
+        rocksdb: RocksDB<StateKeeperColumnFamily>,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        tracing::debug!("Catching up RocksDB synchronously");
+        let rocksdb_builder: RocksdbStorageBuilder = rocksdb.into();
+        let rocksdb = rocksdb_builder
+            .synchronize(conn, stop_receiver)
+            .await
+            .context("Failed to catch up state keeper RocksDB storage to Postgres")?;
+        let Some(rocksdb) = rocksdb else {
+            tracing::info!("Synchronizing RocksDB interrupted");
+            return Ok(None);
+        };
+        let rocksdb_l1_batch_number = rocksdb.l1_batch_number().await.unwrap_or_default();
+        tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
+        Ok(Some(rocksdb.into()))
+    }
+
+    async fn access_storage_inner<'a>(
+        &'a self,
+        rt_handle: Handle,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        match self.rocksdb_cell.get() {
+            Some(rocksdb) => {
```

Comment:
Bikshedding: `if let Some(_) = self.rocksdb_cell.get() { .. } else { .. }` is slightly shorter and sometimes considered more idiomatic.